Issue #14929: added support for violation messages for google tests

Zopsss opened the pull request:
issue #14929 

added support for violations messages including above, below, etc in AbstractItModuleTestSupport for google IT tests

romani commented:
Please address inspections https://app.circleci.com/pipelines/github/checkstyle/checkstyle/25759/workflows/c168ad05-454d-4f52-ad07-24856fa99a62/jobs/600065/artifacts

romani commented:
Can you migrate at least one test class to this form?

There are bunch of copypasta is there a way to reuse functionality.

Hint: we can update "warn" to "violation" in input files of ITs.

Zopsss commented:
> Can you migrate at least one test class to this form?

Ok I'll update the PR to make any one test class to use this approach.


> There are bunch of copypasta is there a way to reuse functionality.

Most of the copy/pasted methods are from `AbstractModuleTestSupport` file, these methods were private and not accessible in `AbstractItModuleTestSupport` so I had to copy paste them, if you have any suggestions about how we can reduce it then please let me know.

> Hint: we can update "warn" to "violation" in input files of ITs. 

Yeah we can now also use violation above, below, etc too

romani commented:
Keep in mind we can search replace warn to violation, to simplify migration.

Zopsss commented:
Updated test class with usage of latest method: 

https://github.com/checkstyle/checkstyle/pull/14959/files#diff-af5b305c42978b4abca26356aabfd39d78a647d90bce1c4084fe52ddb066c7b5

https://github.com/checkstyle/checkstyle/pull/14959/files#diff-67db227826a197f750774d7adac71e3df8a855480e4ef371a27e0834bd9cc7e1

https://github.com/checkstyle/checkstyle/pull/14959/files#diff-5a589bc235dd0cf5bba6f60eb87712b5f067a2f65b22aab6e508837888727533


I have used 3 types of violation msgs:

- violation 'msg....'
https://github.com/checkstyle/checkstyle/blob/9360375e1489ea3c052612dc464b3c2ad1b8c2c0/src/it/resources/com/google/checkstyle/test/chapter3filestructure/rule3421overloadsplit/InputConstructorsDeclarationGrouping.java#L39
- violation above 'msg....'
https://github.com/checkstyle/checkstyle/blob/9360375e1489ea3c052612dc464b3c2ad1b8c2c0/src/it/resources/com/google/checkstyle/test/chapter3filestructure/rule3421overloadsplit/InputConstructorsDeclarationGrouping.java#L17-L18
- violation below 'msg....'
https://github.com/checkstyle/checkstyle/blob/9360375e1489ea3c052612dc464b3c2ad1b8c2c0/src/it/resources/com/google/checkstyle/test/chapter3filestructure/rule3421overloadsplit/InputConstructorsDeclarationGrouping.java#L82-L83

Though we have support of all kinds of messages.

---

In the current method, we need to compulsory use the `// violation` with its message even if there is only one module we're loading. If you want, I can modify the method to enforce using messages only when we're loading more than one module. Let me know what do you think about this

Zopsss commented:
The [circlecli: run-inspection](https://circleci-tasks-prod.s3.us-east-1.amazonaws.com/storage/artifacts/2233fdd7-79cb-481c-a64c-0df373c0d898/c2c5562c-cb2d-443b-bcfd-8a51a7f80c70/0/home/circleci/project/target/inspection-results/IfStatementWithTooManyBranches.xml?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=ASIAQVFQINEOPD5Q7KHK%2F20240610%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20240610T191958Z&X-Amz-Expires=60&X-Amz-Security-Token=IQoJb3JpZ2luX2VjELT%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLWVhc3QtMSJHMEUCIEQg2J3ZcMbqZjBxtSEonttVQXXi7InUB%2F99ejf%2FPZKNAiEA1QRdCXaY59u6gtVTK1SE0c9LsIJ%2Besw%2BGhpBmKjGftsqqwIITBADGgwwNDU0NjY4MDY1NTYiDFwG7ICL3NC%2FIwosmSqIAnbJkToCzTsDieetFiwc1tAKniD9tYPo11f%2FHcuKipPrNAmv%2FRMY1k4%2B9c481LSaFUTjLpR22AW%2FgQa%2FqNr4jRphAF3tCS9NRBsjBxq6XdCRt3wSFE7M2WzI0AEkkYuFvveGSoyvszn4z4Kx7TluNZIE5N0SGVbMWoB6mgJ3DYFBK%2Bkt%2B6XM8DsI%2BwYUE0TDHMdMU7A6sJm9MW%2BnvEpkbWKyocoVAyzXRjH%2BczOh%2BV6FiniZTE0mGzivlNJVK7n1%2BdZumuSAV8pIdmQwuvSCqOIuSO8NfWRvS3sVPCOZ3LMZJLRF5%2FCM7CqWcUfRHyXM%2FOgSRXziCmuw8rqDc7pfPk1PcSsMXiXcvDDCnZ2zBjqdAUu129STC0dnXq4MNU3nhNtFklK4WqtwNR8V8oqtcKTZnuzzgoHGT%2BP3uhAbM7zqlhndJtqNuDjBuzth7gTlj1J9Jw6fLkn0T1zcN0l4zAzamGjhKgKmFgJthFa3KCwebNi2%2ByRZ8%2FukwQcg2Yi9cjsaityKFfi%2BBHqyh9ixRDX5byoy8Zg%2FniZe6LI%2BG9pjxsbKoF1HilCyQHzbCLs%3D&X-Amz-SignedHeaders=host&x-id=GetObject&X-Amz-Signature=9e7604c30412a8acaf9db4bead720b6a334d156f6a1be638c5c78ffd986f4e05) failing CI is for following lines:

https://github.com/checkstyle/checkstyle/blob/9360375e1489ea3c052612dc464b3c2ad1b8c2c0/src/test/java/com/puppycrawl/tools/checkstyle/bdd/InlineConfigParser.java#L723-L730

it is a long list of if-else-if ladder. This is copied directly from with little modifications: 

https://github.com/checkstyle/checkstyle/blob/9360375e1489ea3c052612dc464b3c2ad1b8c2c0/src/test/java/com/puppycrawl/tools/checkstyle/bdd/InlineConfigParser.java#L548-L550

It plays an important role to detect the `// violation` msg type ( like above, below, etc ). I'm not sure how we can resolve this error.

rdiachenko commented:
> I'm not sure how we can resolve this error.

One way to get rid of that many if-else conditions is to refactor it by grouping patterns and their corresponding conditional logic together. Here's idea for you how it could be done, please play with the code, do change it:
```java
// Construct a map of pattern and its corresponding logic.
// `LinkedHashMap` allows to preserve ordering.
final Map<Pattern, Consumer<Matcher>> violationPatternMatchers = new LinkedHashMap<>();
        violationPatternMatchers.put(VIOLATION_PATTERN, m -> {
            final String violationMessage = m.group(1);
            final int violationLineNum = lineNo + 1;
            checkWhetherViolationSpecified(specifyViolationMessage, violationMessage,
                    violationLineNum);
            inputConfigBuilder.addViolation(violationLineNum, violationMessage);
        });

...

violationPatternMatchers.put(VIOLATIONS_SOME_LINES_ABOVE_PATTERN, m -> {
            Collections
                    .nCopies(Integer.parseInt(m.group(1)),
                            lineNo + 2)
                    .forEach(actualLineNumber -> {
                        inputConfigBuilder.addViolation(actualLineNumber, null);
                    });
        });

// Find the first matcher which matches the pattern and apply it.
// Otherwise, go with default logic which uses `useFilteredViolations`.
violationPatternMatchers.keySet().stream()
                .filter(p -> p.matcher(lines.get(lineNo)).matches())
                .findFirst()
                .ifPresentOrElse(p -> {
                    violationPatternMatchers.get(p).accept(p.matcher(lines.get(lineNo)));
                }, () -> {
                    if (useFilteredViolations) {
                        setFilteredViolation(inputConfigBuilder, lineNo + 1,
                                lines.get(lineNo), specifyViolationMessage);
                    }
        });
```

Zopsss commented:
> One way to get rid of that many if-else conditions is to refactor it by grouping patterns and their corresponding conditional logic together. 

I have made changes according to your suggestions, just had to modify the code a little bit differently from what you provided. But I'm facing following checkstyle errors:

```
[ERROR] [checkstyle] [ERROR] C:\OpenSource\checkstyle\src\test\java\com\puppycrawl\tools\checkstyle\bdd\InlineConfigParser.java:698:5: Cyclomatic Complexity is 22 (max allowed is 10). [CyclomaticComplexity]
[ERROR] [checkstyle] [ERROR] C:\OpenSource\checkstyle\src\test\java\com\puppycrawl\tools\checkstyle\bdd\InlineConfigParser.java:698:5: NCSS for this method is 90 (max allowed is 50). [JavaNCSS]
[ERROR] [checkstyle] [ERROR] C:\OpenSource\checkstyle\src\test\java\com\puppycrawl\tools\checkstyle\bdd\InlineConfigParser.java:707:65: Lambda body length is 14 lines (max allowed is 10). [LambdaBodyLength]
[ERROR] [checkstyle] [ERROR] C:\OpenSource\checkstyle\src\test\java\com\puppycrawl\tools\checkstyle\bdd\InlineConfigParser.java:722:71: Lambda body length is 12 lines (max allowed is 10). [LambdaBodyLength]
[ERROR] [checkstyle] [ERROR] C:\OpenSource\checkstyle\src\test\java\com\puppycrawl\tools\checkstyle\bdd\InlineConfigParser.java:735:71: Lambda body length is 14 lines (max allowed is 10). [LambdaBodyLength]
[ERROR] [checkstyle] [ERROR] C:\OpenSource\checkstyle\src\test\java\com\puppycrawl\tools\checkstyle\bdd\InlineConfigParser.java:750:88: Lambda body length is 12 lines (max allowed is 10). [LambdaBodyLength]
[ERROR] [checkstyle] [ERROR] C:\OpenSource\checkstyle\src\test\java\com\puppycrawl\tools\checkstyle\bdd\InlineConfigParser.java:763:88: Lambda body length is 14 lines (max allowed is 10). [LambdaBodyLength]
[ERROR] [checkstyle] [ERROR] C:\OpenSource\checkstyle\src\test\java\com\puppycrawl\tools\checkstyle\bdd\InlineConfigParser.java:778:82: Lambda body length is 14 lines (max allowed is 10). [LambdaBodyLength]
[ERROR] [checkstyle] [ERROR] C:\OpenSource\checkstyle\src\test\java\com\puppycrawl\tools\checkstyle\bdd\InlineConfigParser.java:793:82: Lambda body length is 15 lines (max allowed is 10). [LambdaBodyLength]
[ERROR] [checkstyle] [ERROR] C:\OpenSource\checkstyle\src\test\java\com\puppycrawl\tools\checkstyle\bdd\InlineConfigParser.java:809:82: Lambda body length is 15 lines (max allowed is 10). [LambdaBodyLength]
```

I tried to suppress them here:

https://github.com/checkstyle/checkstyle/blob/6205b1067367128812c9bedcddf41a3f49f2c969/src/test/java/com/puppycrawl/tools/checkstyle/bdd/InlineConfigParser.java#L693-L697

But it is still not working. The suppression for `LambdaBodyLength` may be wrong, I also tried to find if can we suppress it via some suppression.xml file but couldn't find any example of it. But the suppression for `CyclomaticComplexity` & `JavaNCSS` is correct. I used the same suppression in the old code and it was working fine. I'm not sure why it is showing error for this time.

rdiachenko commented:
@Zopsss , considering this is a test class and complex refactoring, which may not make this test method more readable, I suggest reverting to `if-else` and suppressing the warning.
@romani agree?

rnveach commented:
> I tried to suppress them here:
> But it is still not working.
> -@cs[LambdaBodyLength] splitting lambdas is not reasonable. 

https://github.com/checkstyle/checkstyle/blob/master/config/checkstyle-checks.xml#L625-L627
`-@cs` suppression has a 3 line reach, meaning the suppression text has to be within 3 lines of the violation. Outside of that and it won't suppress the violation.

> I also tried to find if can we suppress it via some suppression.xml file but couldn't find any example of it

https://github.com/checkstyle/checkstyle/blob/master/config/suppressions.xml#L19
Just swap check and file name with the check and file you are going against.

You would need to verify which config is being run that is generating the violation. If it is the non-main config (however I am not sure if this is the correct config), then the suppression needs to be added to https://github.com/checkstyle/checkstyle/blob/master/config/checkstyle-non-main-files-suppressions.xml .

Zopsss commented:
> I suggest reverting to if-else and suppressing the warning.

done

Zopsss commented:
> > I tried to suppress them here:
> > But it is still not working.
> > `-@cs[LambdaBodyLength] splitting lambdas is not reasonable. `
> 
> https://github.com/checkstyle/checkstyle/blob/master/config/checkstyle-checks.xml#L625-L627 `-@cs` suppression has a 3 line reach, meaning the suppression text has to be within 3 lines of the violation. Outside of that and it won't suppress the violation.
> 
> > I also tried to find if can we suppress it via some suppression.xml file but couldn't find any example of it
> 
> https://github.com/checkstyle/checkstyle/blob/master/config/suppressions.xml#L19 Just swap check and file name with the check and file you are going against.
> 
> You would need to verify which config is being run that is generating the violation. If it is the non-main config (however I am not sure if this is the correct config), then the suppression needs to be added to https://github.com/checkstyle/checkstyle/blob/master/config/checkstyle-non-main-files-suppressions.xml . The config being used also plays in if the `-@cs` suppression will work or not, as if its not defined in the config, then you can't use that form of suppression.

Thank you for the help, I tried this and the suppressions were working fine, the build was passing locally. But as Ruslan suggested in few comments above, I reverted the changes and added missing suppressions now all the CIs are passing.

romani commented on src/test/java/com/puppycrawl/tools/checkstyle/bdd/InlineConfigParser.java:
```diff
@@ -667,6 +667,152 @@ else if (useFilteredViolations) {
+    // -@cs[JavaNCSS] splitting this method is not reasonable.
+    // -@cs[CyclomaticComplexity] splitting this method is not reasonable.
+    private static List<TestInputViolation> getViolationsForGoogleConfig(List<String> lines,
+         int lineNo) throws CheckstyleException {
```
copy of this is concern.

can we reuse:
```
private static void setViolations(TestInputConfiguration.Builder inputConfigBuilder,
                                      List<String> lines, boolean useFilteredViolations,
                                      int lineNo, boolean specifyViolationMessage)
            throws CheckstyleException {
```

you use useFilteredViolations as true
and we just need to go inputConfigBuilder.getViolations to get only list of violations.
In some future we will refactor this code at bit better.

Zopsss replied:
Done. Thanks for the advice, this helped a lot to reduce redundant code

romani reviewed (changes requested):
just for me:
```
public class ConstructorsDeclarationGroupingTest extends AbstractGoogleModuleTestSupport 
abstract class AbstractGoogleModuleTestSupport extends AbstractItModuleTestSupport
public abstract class AbstractItModuleTestSupport extends AbstractPathTestSupport {
public abstract class AbstractPathTestSupport {



public class ConstructorsDeclarationGroupingCheckTest extends AbstractModuleTestSupport {
public abstract class AbstractModuleTestSupport extends AbstractPathTestSupport {
public abstract class AbstractPathTestSupport {
```

AbstractItModuleTestSupport and AbstractModuleTestSupport are very different. Refactoring will not be simple.

items:

Zopsss commented:
> AbstractItModuleTestSupport and AbstractModuleTestSupport are very different. Refactoring will not be simple.

I required to copy some methods from `AbstractModuleTestSupport`, it was simple copy/pasting and the other things were not too complicated. I guess it will be not much difficult. Your advice helped a lot to reuse the existing code

Zopsss commented:
I have changed another test ( ClassAnnotationsTest ) to use this method:

https://github.com/checkstyle/checkstyle/blob/d3ada8a82e0d7892eeb1bb80fd2ced30826409a0/src/it/java/com/google/checkstyle/test/chapter4formatting/rule4852classannotations/ClassAnnotationsTest.java#L37-L45

This test combines 2 different modules: `AnnotationLocation` & `InvalidJavadocPosition`

In the following line:

https://github.com/checkstyle/checkstyle/blob/d3ada8a82e0d7892eeb1bb80fd2ced30826409a0/src/it/java/com/google/checkstyle/test/chapter4formatting/rule4852classannotations/ClassAnnotationsTest.java#L39

The `new String[] {"AnnotationLocationMostCases"}` is the moduleId, I have kept it String array so if we want we can fetch the same module multiple times with different moduleId.

romani commented on src/it/java/org/checkstyle/base/AbstractItModuleTestSupport.java:
```diff
@@ -451,6 +481,200 @@ protected final void verify(Checker checker,
+            for (String moduleId : moduleIds) {
+                final Configuration moduleConfig =
+                        getModuleConfig(masterGoogleConfig, entry.getKey(), moduleId);
+                root.addChild(moduleConfig);
```
please move this operation outside of method, no reason to provide root as param, and rename method to getCheckerChildren

romani replied:
same for handleTreeWalkerChildren

Zopsss replied:
Done. Here's the new changes:

https://github.com/checkstyle/checkstyle/blob/a0f82691156c959fbb7dccd991c3e5c4f83022e1/src/it/java/org/checkstyle/base/AbstractItModuleTestSupport.java#L620-L631

https://github.com/checkstyle/checkstyle/blob/a0f82691156c959fbb7dccd991c3e5c4f83022e1/src/it/java/org/checkstyle/base/AbstractItModuleTestSupport.java#L641-L652

https://github.com/checkstyle/checkstyle/blob/a0f82691156c959fbb7dccd991c3e5c4f83022e1/src/it/java/org/checkstyle/base/AbstractItModuleTestSupport.java#L570-L585

this may be not efficient, I'm not sure if this is the exact changes you requested. Please let me know if I still need to change something.

romani commented on src/it/java/org/checkstyle/base/AbstractItModuleTestSupport.java:
```diff
@@ -451,6 +481,200 @@ protected final void verify(Checker checker,
+     * @param listOfModules list of modules to load from google config.
+     * @return google configuration based on the list of modules.
+     */
+    private static DefaultConfiguration createGoogleConfig(Map<String, String[]> listOfModules) {
```
createConfigForModulesOfGoogleConfig

Zopsss replied:
done.

https://github.com/checkstyle/checkstyle/blob/a0f82691156c959fbb7dccd991c3e5c4f83022e1/src/it/java/org/checkstyle/base/AbstractItModuleTestSupport.java#L560-L564

romani commented on src/it/java/org/checkstyle/base/AbstractItModuleTestSupport.java:
```diff
@@ -451,6 +481,200 @@ protected final void verify(Checker checker,
+    public final void verifyWithGoogleConfigParser(Map<String, String[]> listOfModules,
+           String filePath) throws Exception {
+        final DefaultConfiguration googleConfig = createGoogleConfig(listOfModules);
+        final TestInputConfiguration testInputConfiguration =
+                InlineConfigParser.parseGoogleConfig(filePath);
+        final List<String> actualViolations = getActualViolationsForFile(googleConfig, filePath);
+
+        verifyViolations(filePath, testInputConfiguration.getViolations(), actualViolations);
```
>InlineConfigParser.parseGoogleConfig(filePath);

this is nonsense, we do not parse config out of Input.
please rename it to getViolationsFromIntegrationInput

Zopsss replied:
done

https://github.com/checkstyle/checkstyle/blob/a0f82691156c959fbb7dccd991c3e5c4f83022e1/src/it/java/org/checkstyle/base/AbstractItModuleTestSupport.java#L493-L497

https://github.com/checkstyle/checkstyle/blob/a0f82691156c959fbb7dccd991c3e5c4f83022e1/src/test/java/com/puppycrawl/tools/checkstyle/bdd/InlineConfigParser.java#L221-L224

romani commented on src/it/java/com/google/checkstyle/test/chapter4formatting/rule4852classannotations/ClassAnnotationsTest.java:
```diff
@@ -38,31 +35,12 @@ protected String getPackageLocation() {
-        };
+        final Map<String, String[]> listOfModules = new HashMap<>();
+        listOfModules.put("AnnotationLocation", new String[] {"AnnotationLocationMostCases"});
+        listOfModules.put("InvalidJavadocPosition", null);
```
```
        listOfModules.put("AnnotationLocationMostCases");
        listOfModules.put("InvalidJavadocPosition");
```

id and name of module should be unique in config.
so in handleCheckerChildren you try to load by class name first, if there is exception, you should load by id.

Zopsss replied:
Done. 

https://github.com/checkstyle/checkstyle/blob/a0f82691156c959fbb7dccd991c3e5c4f83022e1/src/it/java/com/google/checkstyle/test/chapter4formatting/rule4852classannotations/ClassAnnotationsTest.java#L35-L42

romani commented on src/test/java/com/puppycrawl/tools/checkstyle/bdd/InlineConfigParser.java:
```diff
@@ -219,6 +219,23 @@ private static TestInputConfiguration parse(String inputFilePath,
+            throw new CheckstyleException(ex.getMessage() + " in " + inputFilePath, ex);
+        }
+
+        return testInputConfigBuilder.build();
```
there is no config in ITs, you should return only violations, this is exactly what you need, and method should do only this.
fact that you reuse TestInputConfiguration.Builder is internal details that shuold not be leaked outside of this class.

Zopsss replied:
done

https://github.com/checkstyle/checkstyle/blob/a0f82691156c959fbb7dccd991c3e5c4f83022e1/src/test/java/com/puppycrawl/tools/checkstyle/bdd/InlineConfigParser.java#L235-L238

romani commented on src/test/java/com/puppycrawl/tools/checkstyle/bdd/InlineConfigParser.java:
```diff
@@ -517,6 +534,14 @@ private static void setProperties(String inputFilePath, Configuration module,
 
+    private static void setViolationsForGoogleConfig(
+            TestInputConfiguration.Builder inputConfigBuilder,
+            List<String> lines) throws CheckstyleException {
```
please inline this method to where it is used, it is too small

Zopsss replied:
done

https://github.com/checkstyle/checkstyle/blob/a0f82691156c959fbb7dccd991c3e5c4f83022e1/src/test/java/com/puppycrawl/tools/checkstyle/bdd/InlineConfigParser.java#L227-L233

romani reviewed (changes requested):
items:

romani commented on src/it/java/org/checkstyle/base/AbstractItModuleTestSupport.java:
```diff
@@ -451,6 +481,200 @@ protected final void verify(Checker checker,
+
+    /**
+     * Returns the actual violations for each file that has been
+     * checked against google config modules.
```
please make javadoc same as source class, it is copy paste and it should be full copy paste that is easy to detect and resolve later on.

Zopsss replied:
done 

https://github.com/checkstyle/checkstyle/blob/a0f82691156c959fbb7dccd991c3e5c4f83022e1/src/it/java/org/checkstyle/base/AbstractItModuleTestSupport.java#L523-L531

romani reviewed (changes requested):
item:

romani reviewed (changes requested):
Last

Zopsss commented on src/test/java/com/puppycrawl/tools/checkstyle/bdd/InlineConfigParser.java:
```diff
@@ -219,6 +219,25 @@ private static TestInputConfiguration parse(String inputFilePath,
         return testInputConfigBuilder.build();
     }
 
+    public static List<TestInputViolation> getViolationsFromIntegrationInput(String inputFilePath)
```
done

Zopsss commented on src/it/java/org/checkstyle/base/AbstractItModuleTestSupport.java:
```diff
@@ -451,6 +480,203 @@ protected final void verify(Checker checker,
+     * @return list of tree walker children.
+     * @throws CheckstyleException if there is a problem loading the configuration.
+     */
+    private static List<Configuration> getTreeWalkerChildren(Configuration masterGoogleConfig,
```
done. new method:

https://github.com/checkstyle/checkstyle/blob/98a990efdc43e905e253d10bc2a7b88b52ee75c3/src/it/java/org/checkstyle/base/AbstractItModuleTestSupport.java#L610-L629

Zopsss commented on src/it/java/org/checkstyle/base/AbstractItModuleTestSupport.java:
```diff
@@ -451,6 +480,203 @@ protected final void verify(Checker checker,
+
+        for (String module : listOfModules) {
+            if (CHECKER_CHILDREN.contains(module)) {
+                final List<Configuration> checkerChildren =
```
done 

https://github.com/checkstyle/checkstyle/blob/98a990efdc43e905e253d10bc2a7b88b52ee75c3/src/it/java/org/checkstyle/base/AbstractItModuleTestSupport.java#L570-L583

Zopsss commented on src/it/java/org/checkstyle/base/AbstractItModuleTestSupport.java:
```diff
@@ -451,6 +480,203 @@ protected final void verify(Checker checker,
+            if (CHECKER_CHILDREN.contains(module)) {
+                final List<Configuration> checkerChildren =
+                        getCheckerChildren(masterGoogleConfig, module);
+                for (Configuration checkerChild : checkerChildren) {
```
done

romani reviewed (approved):
Ok to merge

rdiachenko reviewed (approved):
lgtm
